### PR TITLE
feat: Add Synthetix

### PR DESCRIPTION
**Pull request by hanjano**

### CODE REVIEW

**Review by galvanizze — APPROVED** [review]

File: blockapi/v2/api/synthetix/synthetix.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        address = ensure_checksum_address(address)
+        return list(self.yield_balances(address))
+
+    def yield_balances(self, address) -> Iterable[BalanceItem]:
+        """
+        Fetch all balances for snx and synth tokens.
+        """
+        staking = self.fetch_staking(address)
+
+        logger.info("Called Synthetix.yield_balances for address: %s", address)
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:169]: nitpick: I would expect this log to be on the first line in `yield_balances`...

File: blockapi/v2/api/synthetix/synthetix.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    collateral: Decimal
+    rewards: Rewards
+    staking_tokens: List[StakingToken]
+
+
+class Synth(TypedDict):
+    symbol: str
+    contract_address: str
+
+    pass
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:70]: nitpick: this file is huge, I would probably split it

File: blockapi/v2/api/synthetix/synthetix.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+            'staking_tokens': list(
+                self.yield_token_staking_data(address, staking_tokens)
+            ),
+        }
+
+    def get_snx_transferable_amount(self, address: str) -> Decimal:
+        """
+        Returns total amount of unlocked SNX tokens.
+        """
+        snx_contract = self.w3.eth.contract(
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:253]: nitpick: I see this in multiple methods - what about some lazy loaded instance attribute? edit: ok I see it is cached, I just saw repeating of the same code... changed to nitpick

File: blockapi/v2/api/synthetix/synthetix.py [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+        """
+        Returns total debt owed in sUSD.
+        """
+        snx_contract = self.w3.eth.contract(
+            snx_contract_address('Synthetix', self.network), abi=synthetix_abi
+        )
+        debt = easy_call(snx_contract, 'debtBalanceOf', address, b'sUSD')
+        return to_decimal(debt)
+
+    @lru_cache(maxsize=128)
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:269]: How did you choose which methods will be cached and which not?

File: blockapi/v2/api/synthetix/synthetix.py [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+                "exchange": to_decimal(fees[1][0]),
+                "staking": to_decimal(fees[1][1]),
+            },
+            "this_week": {
+                "exchange": to_decimal(fees[0][0]),
+                "staking": to_decimal(fees[0][1]),
+            },
+        }
+
+    def get_fee_period_info(self) -> Dict:
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:344]: Will this be used somewhere or is it forgotten?

File: blockapi/v2/api/synthetix/synthetix.py [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+        )
+        total_escrowed = easy_call(
+            rewards_escrow_contract,
+            "totalEscrowedAccountBalance",
+            address,
+        )
+
+        return to_decimal(total_escrowed)
+
+    def get_total_rewards(self, address: str) -> Decimal:
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:419]: will be used?

File: blockapi/v2/api/synthetix/synthetix.py [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+            snx_contract_address('RewardEscrowV2', self.network),
+            abi=rewards_escrow_v2_abi,
+        )
+        total_vested = easy_call(
+            rewards_escrow_contract, "totalVestedAccountBalance", address
+        )
+
+        return to_decimal(total_vested)
+
+    def get_token_xchg_rates(self, synths: List[Synth]) -> Dict:
```
Comment (ladovina) [inline blockapi/v2/api/synthetix/synthetix.py:433]: will be used?

File: blockapi/v2/api/synthetix/synthetix.py (hunk 1 above)
Comment (hanjano) [reply blockapi/v2/api/synthetix/synthetix.py:165]: and so would I  :+1:

File: blockapi/v2/api/synthetix/synthetix.py (hunk 5 above)
Comment (hanjano) [reply blockapi/v2/api/synthetix/synthetix.py]: Removed, thanks.

File: blockapi/v2/api/synthetix/synthetix.py (hunk 6 above)
Comment (hanjano) [reply blockapi/v2/api/synthetix/synthetix.py]: Removed, thanks.

File: blockapi/v2/api/synthetix/synthetix.py (hunk 7 above)
Comment (hanjano) [reply blockapi/v2/api/synthetix/synthetix.py:395]: This might.

**Review by ladovina — APPROVED** [review]

**Review by ladovina — APPROVED** [review]

**Pull request merged**